components/accelerator/nvidia/infiniband/component.go

```diff
@@ -39,92 +43,218 @@ func SetDefaultExpectedPortStates(states infiniband.ExpectedPortStates) {
 	defaultExpectedPortStates = states
 }
 
-func New(ctx context.Context, toolOverwrites nvidia_common.ToolOverwrites) components.Component {
-	return &component{
+func New(ctx context.Context, dbRW *sql.DB, dbRO *sql.DB, toolOverwrites nvidia_common.ToolOverwrites) (components.Component, error) {
+	eventsStore, err := events_db.NewStore(
+		dbRW,
+		dbRO,
+		events_db.CreateDefaultTableName(nvidia_infiniband_id.Name),
+		3*24*time.Hour,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	cctx, ccancel := context.WithCancel(ctx)
+	c := &component{
+		rootCtx:        cctx,
+		cancel:         ccancel,
+		eventsStore:    eventsStore,
 		toolOverwrites: toolOverwrites,
 	}
+
+	return c, nil
 }
 
 var _ components.Component = (*component)(nil)
 
 type component struct {
+	rootCtx        context.Context
+	cancel         context.CancelFunc
+	eventsStore    events_db.Store
 	toolOverwrites nvidia_common.ToolOverwrites
+
+	lastEventMu sync.RWMutex
+	lastEvent   *components.Event
+}
+
+func (c *component) getLastEvent() *components.Event {
+	c.lastEventMu.RLock()
+	defer c.lastEventMu.RUnlock()
+	return c.lastEvent
+}
+
+func convertToState(ev *components.Event) components.State {
+	state := components.State{
+		Name:             ev.Name,
+		Healthy:          true,
+		Health:           components.StateHealthy,
+		Reason:           ev.Message,
+		SuggestedActions: ev.SuggestedActions,
+	}
+	if len(ev.ExtraInfo) > 0 {
+		state.Healthy = ev.ExtraInfo["state_healthy"] == "true"
+		state.Health = ev.ExtraInfo["state_health"]
+	}
+	return state
+}
+
+func (c *component) checkIbstatOnce(ts time.Time, thresholds infiniband.ExpectedPortStates) error {
+	if thresholds.AtLeastPorts <= 0 && thresholds.AtLeastRate <= 0 {
+		return nil
+	}
+
+	ev := components.Event{
+		Time:    metav1.Time{Time: ts},
+		Name:    "ibstat",
+		Type:    common.EventTypeUnknown,
+		Message: "",
+		ExtraInfo: map[string]string{
+			"state_healthy": "true",
+			"state_health":  components.StateHealthy,
+		},
+		SuggestedActions: nil,
+	}
+
+	c.lastEventMu.Lock()
+	defer c.lastEventMu.Unlock()
+
+	cctx, ccancel := context.WithTimeout(c.rootCtx, 15*time.Second)
```

Comment:
We should check if the duration from last call this function to now exceeds a threshold, and return the last event if it does not exceed the threshold.

Reply:
Oh it's done outside of this function

```go
	// last event already happened within the last 10 seconds, skip the check
	if lastEvent != nil && now.UTC().Sub(lastEvent.Time.Time) < 10*time.Second {
		return []components.State{convertToState(lastEvent)}, nil
	}

	if err := c.checkIbstatOnce(now.UTC(), thresholds); err != nil {
		return nil, err
	}
```

Reply:
let me put that logic inside